storage/s3.go

```diff
@@ -62,29 +62,43 @@ type S3 struct {
 	partSize          int64
 }
 
-type s3Opt func(s3 *S3)
+type S3Opt func(s3 *S3) error
 
 // WithRetryPolicy is an option to constructor NewS3 to add a Retry Policy
 // impacting GET operations
-func WithRetryPolicy(policy RetryPolicy) s3Opt {
-	return s3Opt(func(s3 *S3) {
+func WithRetryPolicy(policy RetryPolicy) S3Opt {
+	return S3Opt(func(s3 *S3) error {
 		s3.retryPolicy = policy
+		return nil
 	})
 }
 
-func WithPartSize(size int64) s3Opt {
-	return s3Opt(func(s3 *S3) {
+func WithPartSize(size int64) S3Opt {
+	return S3Opt(func(s3 *S3) error {
 		s3.partSize = size
+		return nil
 	})
 }
 
-func WithUploadConcurrency(concurrency int) s3Opt {
-	return s3Opt(func(s3 *S3) {
+func WithUploadConcurrency(concurrency int) S3Opt {
+	return S3Opt(func(s3 *S3) error {
 		s3.uploadConcurrency = concurrency
+		return nil
+	})
+}
+
+func WithAllowList(l []string) S3Opt {
+	return S3Opt(func(s3 *S3) error {
+		for _, url := range l {
+			if strings.HasPrefix(s3.cfg.Endpoint, url) {
+				return nil
+			}
+		}
+		return errors.New("endpoint is not in allow list")
 	})
 }
 
-func NewS3(cfg S3Config, opts ...s3Opt) *S3 {
+func NewS3(cfg S3Config, opts ...S3Opt) (*S3, error) {
```

Comment:
Issue: Usually we don't want `New` functions to return error. It should only instantiate the object.
In addition, this will break retro-compatibility.

Suggestion: From https://github.com/Scalingo/go-utils/issues/804#issuecomment-2024912293 suggestion is to log something in run time.
But if you really prefer to throw an error, I'd see either:
- return it in runtime (e.g. `Get`, `Upload`, ...) but not very practical in fact
- create a dedicated helper to validate the list. But it need to add this check everywhere.
- keep it like this. In that case take look at the current implementations in our code base

Reply:
I changed the way it's handled. WDYT?

Reply:
Sorry for the delay and the change of mind, as we are in a context of configuration issue we should get an error as soon as possible. Hence the need to use an error in constructor (so your proposal in this commit: https://github.com/Scalingo/go-utils/compare/d599d4eee09fceb8c27e60c30dff5fa7b5a238dc...fddf2c502859a62b128e658aa53fd0287a7f6ad5) :+1:

From our implementation in code base (e.g. [on database-core](https://github.com/Scalingo/appsdeck-database/blob/e3964de87b54eb736b20d9b74db1a9fbe0d79047/controller/backup/storage/s3/s3.go#L30-L66)) we should make some change to fit the logic, i.e add a `BackendGetter` struct on boot of the related service.

Note: this will imply to make a major update of the package.

For the record: https://scalingo.slack.com/archives/CRC7L0XM2/p1714063053317099?thread_ts=1714058866.915739&cid=CRC7L0XM2